source/Irrlicht/CIrrDeviceSDL.cpp

```diff
@@ -105,6 +107,91 @@ EM_BOOL CIrrDeviceSDL::MouseLeaveCallback(int eventType, const EmscriptenMouseEv
 }
 #endif
 
+
+bool CIrrDeviceSDL::keyIsKnownSpecial(EKEY_CODE key) const
+{
+	switch ( key )
+	{
+		// keys which are known to have safe special character interpretation
+		// could need changes over time (removals and additions!)
+		case KEY_RETURN:
+		case KEY_PAUSE:
+		case KEY_ESCAPE:
+		case KEY_PRIOR:
+		case KEY_NEXT:
+		case KEY_HOME:
+		case KEY_END:
+		case KEY_LEFT:
+		case KEY_UP:
+		case KEY_RIGHT:
+		case KEY_DOWN:
+		case KEY_PRINT:
+		case KEY_SNAPSHOT:
+		case KEY_INSERT:
+		case KEY_BACK:
+		case KEY_DELETE:
+		case KEY_HELP:
+		case KEY_APPS:
+		case KEY_SLEEP:
+		case KEY_F1:
+		case KEY_F2:
+		case KEY_F3:
+		case KEY_F4:
+		case KEY_F5:
+		case KEY_F6:
+		case KEY_F7:
+		case KEY_F8:
+		case KEY_F9:
+		case KEY_F10:
+		case KEY_F11:
+		case KEY_F12:
+		case KEY_F13:
+		case KEY_F14:
+		case KEY_F15:
+		case KEY_F16:
+		case KEY_F17:
+		case KEY_F18:
+		case KEY_F19:
+		case KEY_F20:
+		case KEY_F21:
+		case KEY_F22:
+		case KEY_F23:
+		case KEY_F24:
+		case KEY_NUMLOCK:
+		case KEY_SCROLL:
+		case KEY_LCONTROL:
+		case KEY_RCONTROL:
+			return true;
+
+		default:
+			return false;
+	}
+}
+
+int CIrrDeviceSDL::findCharToPassToIrrlicht(int assumedChar, EKEY_CODE key) {
+	switch (key) {
+		case KEY_PRIOR:
+		case KEY_NEXT:
+		case KEY_HOME:
+		case KEY_END:
+		case KEY_LEFT:
+		case KEY_UP:
+		case KEY_RIGHT:
+		case KEY_DOWN:
+			return 0;
+		default:
+			return assumedChar;
+	}
+}
+
+void CIrrDeviceSDL::resetReceiveTextInputEvents() {
+	gui::IGUIElement *elem = GUIEnvironment->getFocus();
+	if (elem && elem->acceptsIME())

```

Comment:
Same suggestion as already mentioned multiple times.
Add a function in the driver that the gui code can call to enter/leave text mode. And make behaviour dependent on text mode, and do not access gui stuff from here.

Reply:
So call this in minetest/minetest? Was that what you meant earlier?

Reply:
So create a function "enable/disable input events" (which calls the relevant SDL toggle) and then call that in minetest/minetest when input mode changes, is that what you mean?

Reply:
Yes. But note that `field[]` uses the `CGUIEditBox` from irrlicht, so not only call it from minetest.

Reply:
I've condensed that part of the code down to a function call, receiveTextInputEvents, but I'm not sure where I'd need to call that. Any suggestions?

Reply:
* When creating the driver (text mode is on by default according to some sdl issue).
* For cguieditbox and guieditbox in minetest in the event handler, there should be events for focus gain and lost, see e.g. here: https://github.com/minetest/irrlicht/blob/29a448de4d16ed4360e31e3e71dd79a269f52404/source/Irrlicht/CGUIEditBox.cpp#L238
* In minetest in the chat console thing when it's opened / closed.

(Btw. I should mention that you don't *have* to follow my suggestions, as I'm not a coredev. I just suggested this again because it seemed to me like you didn't quite understand me.)

Reply:
The change does sound sensible, it would be good to separate out focus/unfocus from the SDL event loop. I was a bit worried that it would be quite hard to find exactly all the places in the code where focus is gained and lost. (Granted, my tool of choice is grep so it might just be that.)

FWIW, CIrrDeviceLinux.cpp (X11) runs acceptsIME directly as well.

I'll give this a few more days to see if I can find them and figure it out but if not I think I'll stick with the original code, if that's fine with you. If you decide to write one, I am happy to take up a direct patch as well.